protoarray/src/main/java/tech.pegasys.artemis.protoarray/ProtoArray.java

```diff
@@ -0,0 +1,380 @@
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.protoarray;
+
+import com.google.common.primitives.UnsignedLong;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import org.apache.tuweni.bytes.Bytes32;
+
+public class ProtoArray {
+
+  private int pruneThreshold;
+  private UnsignedLong justifiedEpoch;
+  private UnsignedLong finalizedEpoch;
+
+  private final List<ProtoNode> nodes;
+  private final Map<Bytes32, Integer> indices;
+
+  public ProtoArray(
+      int pruneThreshold,
+      UnsignedLong justifiedEpoch,
+      UnsignedLong finalizedEpoch,
+      List<ProtoNode> nodes,
+      Map<Bytes32, Integer> indices) {
+    this.pruneThreshold = pruneThreshold;
+    this.justifiedEpoch = justifiedEpoch;
+    this.finalizedEpoch = finalizedEpoch;
+    this.nodes = nodes;
+    this.indices = indices;
+  }
+
+  public Map<Bytes32, Integer> getIndices() {
+    return indices;
+  }
+
+  public List<ProtoNode> getNodes() {
+    return nodes;
+  }
+
+  public void setPruneThreshold(int pruneThreshold) {
+    this.pruneThreshold = pruneThreshold;
+  }
+
+
+
+  /**
+   * Register a block with the fork choice.
+   * It is only sane to supply a `None` parent for the genesis block.
+   *
+   * @param slot
+   * @param root
+   * @param optionalParentRoot
+   * @param stateRoot
+   * @param justifiedEpoch
+   * @param finalizedEpoch
+   */
+  public void onBlock(
+      UnsignedLong slot,
+      Bytes32 root,
+      Optional<Bytes32> optionalParentRoot,
+      Bytes32 stateRoot,
+      UnsignedLong justifiedEpoch,
+      UnsignedLong finalizedEpoch) {
+    if (indices.containsKey(root)) {
+      return;
+    }
+
+    int nodeIndex = nodes.size();
+
+    ProtoNode node =
+        new ProtoNode(
+            slot,
+            stateRoot,
+            root,
+            optionalParentRoot.map(indices::get),
+            justifiedEpoch,
+            finalizedEpoch,
+            UnsignedLong.ZERO,
+            Optional.empty(),
+            Optional.empty());
+
+    indices.put(node.getRoot(), nodeIndex);
+    nodes.add(node);
+
+    node.getParentIndex()
+        .ifPresent(parentIndex -> maybeUpdateBestChildAndDescendant(parentIndex, nodeIndex));
+  }
+
+  /**
+   * Follows the best-descendant links to find the best-block (i.e., head-block).
+   *
+   * The result of this function is not guaranteed to be accurate if `onBlock` has
+   * been called without a subsequent `applyScoreChanges` call. This is because
+   * `onBlock` does not attempt to walk backwards through the tree and update the
+   * best child / best descendant links.
+   *
+   * @param justifiedRoot
+   * @return
+   */
+  public Bytes32 findHead(Bytes32 justifiedRoot) {
+    int justifiedIndex = indices.get(justifiedRoot);
```

Comment:
Probably should check for null here because it confuses the heck out of people when they get a `NullPointerException` from an `int` (it's because `indices` returns an `Integer` which can be `null` and throws `NullPointerException` when unboxed to `int`).
```suggestion
    int justifiedIndex = checkNotNull(indices.get(justifiedRoot), "Unknown justified root");
```

Reply:
Good catch. Added checkNotNull's for both justifiedRoot and justifiedIndex.